granulate_utils/logging/__init__.py

```diff
@@ -0,0 +1,66 @@
+#
+# Copyright (c) Granulate. All rights reserved.
+# Licensed under the AGPL3 License. See LICENSE.md in the project root for license information.
+#
+import logging
+from datetime import datetime
+from logging import Handler, LogRecord
+from typing import Any
+
+
+class BatchingHandler(Handler):
+    """
+    logging.Handler that accumulates logs in a buffer and flushes them upon explicit request.
+    The logs are transformed into dicts.
+    """
+    capacity = 100 * 1000  # max number of records to buffer
+
+    def __init__(self) -> None:
+        super().__init__(logging.DEBUG)
+        self._buffer: list[dict] = []
+        # The formatter is needed to format tracebacks
+        self.setFormatter(logging.Formatter())
+
+    def emit(self, record: LogRecord) -> None:
+        item = self.dictify_record(record)
+        self._buffer.append(item)
+        # truncate logs to last N entries
+        if len(self._buffer) > self.capacity:
+            self.on_truncated()
+            self._buffer[: -self.capacity] = []
```

Comment:
This removes the first N extra items, right? This means `on_truncated` will be called very often if there are errors. I think an API like the way the gProfiler class works makes more sense (i.e pass `was_truncated` when calling `flush_logs`. Or call `on_truncated()` once per `flush_logs` call?

Reply:
Hmm... yes, as a matter of fact most of the time (ie. non-concurrent calls) this would probably remove just one item from the head of the list consistently until flushed. A better option would probably be to toss $k \times Capacity$ on overflow.